java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,354 @@
+  private final static int MAX_CNT_VALUE = 255;
+  private final static int MAX_TAG_LENGTH = 32;
+  private final static int MIN_TAG_LENGTH = 1;
+
+  /**
+   * Generates a pair of private/public keys for the user, calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum allowed peer's tag length.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  public Glome(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.userKeys = generateKeys();
+    initMacKeys(peerKey, minPeerTagLength);
+  }
+
+  /**
+   * Derives the user's public key from the {@code privateKey}, calculates user's and peer's MAC
+   * keys.
+   *
+   * @param peerKey peer's public key.
+   * @param userPrivateKey user's private key.
+   * @param minPeerTagLength minimum allowed peer's tag length.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  public Glome(XECPublicKey peerKey, XECPrivateKey userPrivateKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    XECPublicKey userPublicKey = derivePublicKeyFromPrivate(userPrivateKey);
+    this.userKeys = new KeyPair(userPublicKey, userPrivateKey);
+    initMacKeys(peerKey, minPeerTagLength);
+  }
+
+  public KeyPair userKeys() {
+    return userKeys;
+  }
+
+  public XECPublicKey peerKey() {
+    return peerKey;
+  }
+
+  public XECPublicKey userPublicKey() {
+    return (XECPublicKey) userKeys.getPublic();
+  }
+
+  public XECPrivateKey userPrivateKey() {
+    return (XECPrivateKey) userKeys.getPrivate();
+  }
+
+  /**
+   * Generates a user's tag corresponding to the given message {@code msg} and the counter {@code
+   * cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the peer to the user.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
```

Comment:
peerTag.length will NPE if peerTag is null.

Reply:
Yes, and I could get a lot of NPEs by receiving null objects. What is the best practice for that? We couldn't check for all the exceptions. I may add @NotNull annotation, but it doesn't seem good. Checking for nulls won't be elegant too.

Reply:
I think annotations like @NotNull are good documentation, at least, if you can use them from the libraries you already use. That's usually more up to the existing style of larger projects, and this one is small so its optional. It's good for low-level libraries not to throw unchecked exceptions on bad input, so I would be as defensive as possible on public library methods; like any API it should try to return "Bad Arguments" etc when used incorrectly.

Reply:
The current practice inside Google seems to be to assume that everything is NotNull unless it is specified as Nullable to avoid noise. But for libraries this is a little more tricky as they might be used from an environment (notably e.g. Android Studio) that wants everything to be tagged with NonNull.

I fully support the comment that unchecked exceptions should not happen on bad input.